### PR TITLE
Improve type annotations for `core.structure`

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- Improve type annotations for `core.structure`
- Fix some `type: ignore[reportPossiblyUnboundVariable]` 

#### A question from me:
What is the difference between `np.ndarray`, `ArrayLike` and `NDArray` for typing?
```python
ArrayLike: type[Buffer] | type[_SupportsArray[dtype[Any]]] | type[_NestedSequence[_SupportsArray[dtype[Any]]]] | type[bool] | type[int] | type[float] | type[complex] | type[str] | type[bytes] | type[_NestedSequence[bool | int | float | complex | str | bytes]]

NDArray: type[ndarray[Any, dtype[_ScalarType_co@NDArray]]]
```

### CODE REVIEW

File: pymatgen/core/structure.py [hunk 1]
```diff
@@ -1234,7 +1251,7 @@ def from_spacegroup(
         coords_are_cartesian: bool = False,
         tol: float = 1e-5,
         labels: Sequence[str | None] | None = None,
-    ) -> IStructure | Structure:
```
Comment (DanielYang59) [inline pymatgen/core/structure.py:1237]: Noticed a lot of cases where `IStructure | Structure` is used where there should be `IStructure` (similar for `IMolecule`). Need to look closer to make sure this is not intended.

**Comment by DanielYang59** [comment]
@janosh Can you please review this? Thanks.

**Review by janosh — APPROVED** [review]
thanks @DanielYang59! 👍

**Comment by DanielYang59** [comment]
Thanks for reviewing and the tweaks!

**Comment by wortelboer** [comment]
Hello,
Today I ran into an issue while using pymatgen 2024.5.31, that I think might be related to this contribution. I am using a fresh conda environment, python 3.11, with pymatgen installed via `pip install pymatgen`. I get an error when trying to import from `pymatgen.core.structure`:
```
>>> from pymatgen.core import Structure
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~\AppData\Local\miniconda3\envs\test311\Lib\site-packages\pymatgen\core\__init__.py", line 17, in <module>
    from pymatgen.core.structure import IMolecule, IStructure, Molecule, PeriodicNeighbor, SiteCollection, Structure
  File "~\AppData\Local\miniconda3\envs\test311\Lib\site-packages\pymatgen\core\structure.py", line 39, in <module>
    from typing_extensions import Self
ModuleNotFoundError: No module named 'typing_extensions'
```
If I then install `typing-extensions` separately, things seem to work fine. In v2024.5.31's `setup.py` I see that `typing-extensions` is listed under `extras_require["dev"]`. My understanding of `setup.py` is limited though. Should I install pymatgen in a different way?
Thank you in advance.

**Pull request merged**